spock/src/main/java/org/jboss/weld/spock/impl/EnableWeldInterceptor.java

```diff
@@ -0,0 +1,187 @@
+        return invocation -> {
+            // this is necessary so that a specification scoped interceptor
+            // does not inject into a feature or iteration scoped fixture method call
+            // or that a feature scoped interceptor of a data-driven feature
+            // does not inject into another features' fixture method calls
+            if ((invocation.getFeature() != null) && !handledFeatures.contains(invocation.getFeature())) {
+                invocation.proceed();
+                return;
+            }
+
+            // only continue if there still are missing arguments
+            if (Arrays.stream(invocation.getArguments())
+                    .noneMatch(argument -> argument == MISSING_ARGUMENT)) {
+                invocation.proceed();
+                return;
+            }
+
+            Object testInstance = invocation.getInstance();
+
+            // get the sole weld initiator of this interceptor
+            // in case it is around a specification or parameterized feature
+            WeldInitiator weldInitiator = weldInitiators.get(null);
+            // or get the weld initiator for the current iteration
+            if (weldInitiator == null) {
+                weldInitiator = weldInitiators.get(new IdentityWrapper(testInstance));
+            }
+            // the fixture method interceptors for all features are triggered for each iteration
+            // so if there are multiple features not all have a matching initiator of course
+            if (weldInitiator == null) {
+                invocation.proceed();
+                return;
+            }
+
+            BeanManager beanManager = weldInitiator.getBeanManager();
+            Parameter[] parameters = invocation.getMethod().getReflection().getParameters();
+            Object[] arguments = invocation.getArguments();
+            for (int i = 0; i < arguments.length; i++) {
+                Object argument = arguments[i];
+                if (argument != MISSING_ARGUMENT) {
+                    continue;
+                }
+
+                Parameter parameter = parameters[i];
+                List<Annotation> qualifiers = Arrays
+                        .stream(parameter.getAnnotations())
+                        .filter(annotation -> beanManager.isQualifier(annotation.annotationType()))
+                        .collect(toList());
+
+                if (explicitParamInjection) {
+                    if (qualifiers.isEmpty()) {
+                        continue;
+                    }
+                    arguments[i] = weldInitiator
+                            .select(parameter.getType(), qualifiers.toArray(new Annotation[0]))
+                            .get();
+                } else {
+                    WeldInstance<?> candidates = weldInitiator
+                            .select(parameter.getType(), qualifiers.toArray(new Annotation[0]));
+                    if (candidates.isResolvable()) {
```

Comment:
> This is inconsistent
> Both scenarios should IMO behave equally.

It is not only inconsistent, it is intended. :-D
I don't agree that they should behave the same.
Imho this is how it makes sense the most.
If the user asked for explicit injection and he added a qualifier, the bean must be resolvable or throw an error.
If we are in implicit mode, we try to resolve what we can and leave the rest uninjected, as other extensions running after us could provide a value.

> In case you want to have it the same as JUnit 5 extension, we seem to always try to resolve the bean and we throw exception if there is a failure.

That's actually not the case.
In `resolveParameter` you blindly try to resolve.
But `resolveParameter` is only called if `supportsParameter` returned `true`.
Seeing both methods in combination, explicit&no-qualifiers means no injection,
otherwise only injected if resolvable.

So to be consistent with the JUnit 5 code, I would need to check for `isResolvable` in both cases.
But as I said, if explicit injection is configured and there is a qualifier, imho it should fail.
If you think differently still, I'll change it to also check for resolvable in the explicit case.

Reply:
I'm not sure you read or understood what I wrote in my last comment, because if you did, your whole comment does not make sense to me.

Let me reiterate using a formula:

Current behavior in Jupiter extension:

explicit mode + qualifier present => isResolvable ? inject : leave uninjected
implicit mode => isResolvable ? inject : leave uninjected

Current behavior in Spock extension:

explicit mode + qualifier present => isResolvable ? inject : throw exception
implicit mode => isResolvable ? inject : leave uninjected


At least in the context of Spock "leave uninjected" does not mean you get NPEs, but Spock 2+ will throw an exception immediately if not all parameters were provided by extensions.


You repeatedly claimed now that the Jupiter extension always tries to greedily inject all parameters or throws exception, but as I repeatedly stated this is **not** the case. The Jupiter considers in implicit mode all and in explicit mode ones with qualifier as _potential_ targets, but only ever injects them if they are resolvable or otherwise just do **not** inject them and do **not** show any exception.
And that is what **I** constantly say makes not too much sense to me for explicit + qualifier case.
For the implicit mode it imho has to be the case because other extensions could provide the value so it is bad if this extension too greedily throws an exception.

Reply:
In my personal opinion, if you want it consistent across extensions, then the Jupiter extensions should be made stricter to have the logic I made for the Spock extension. I mainly just didn't adjust this because it would be a potentially breaking change.

Reply:
> You repeatedly claimed now that the Jupiter extension always tries to greedily inject all parameters or throws exception, but as I repeatedly stated this is not the case. The Jupiter considers in implicit mode all and in explicit mode ones with qualifier as potential targets, but only ever injects them if they are resolvable or otherwise just do not inject them and do not show any exception.

I was partly mistaken because I thought in junit it would blow with unsatisfied/ambig dep when you cannot resolve it, so that's totally my bad.
However, what happens is that junit checks for resolvability of all parameters by all extensions and throws its own error when none can resolve it. So you'll end up with `org.junit.jupiter.api.extension.ParameterResolutionException` instead.
The point being, you still cannot start executing a test method with params staying `null` which is what I was after.

Much like what you said for Spock 2:

> At least in the context of Spock "leave uninjected" does not mean you get NPEs, but Spock 2+ will throw an exception immediately if not all parameters were provided by extensions.

I wasn't aware Spock will do the same thing (which is on me, I should have tried it before jumping to assumptions). With that in mind, I think the current impl is correct :)

> And that is what I constantly say makes not too much sense to me for explicit + qualifier case.

This is good point, we could/should make it stricter or maybe log something in case of explicit param injection.
The non-explicit case is IMO correct for junit extension as other extensions can potentially claim those parameters.

Reply:
> Much like what you said for Spock 2:

Actually I think it should even throw for Spock 1, I just don't remember and didn't test. :-D

> The non-explicit case is IMO correct for junit extension as other extensions can potentially claim those parameters.

Exactly, that's why I kept it for Spock too. :-)

> This is good point, we could/should make it stricter or maybe log something in case of explicit param injection.

Do you mean I should change this along with this PR for Jupiter to have it like it is now for Spock?
Or do you mean you are going to merge the PR as is and adapt the Jupiter code later?

Reply:
We can deal with that separately, I'll create a tracking issue.
Later today or tomorrow I'll get to merging the PR as well as looking into a release.

Reply:
Great, the v3 PR is also in mergable state, I did the changes synchronously.
The only difference is what is also different for Jupiter between v3 and `master` in the tests.
And v3 is the release I'm looking for iirc :-)